blog/_src/posts/2017-09-13-redex-faq.md

```diff
@@ -0,0 +1,477 @@
+2. defining _pattern_-based judgments and relations on _terms_
+3. testing properties of the above
+
+More generally, Redex is helpful for experimenting with a programming language
+ design, and helping you decide what you might want to prove about a language.
+
+
+### What is Redex **not** useful for?
+
+Proving theorems about a grammar, judgment, or relation.
+
+
+### What is a _term_?
+
+Informally, a term is:
+
+- a Redex "atom", or
+- an object that represents a sequence of characters.
+
+More formally, a term is the result of evaluating `(term X)`, where `X` is
+ any Racket expression.
+
+Examples:
+
+```
+$ racket
+Welcome to Racket v6.10.0.3.
+> (require redex/reduction-semantics)
+> (term 42)
+42
+> (term (+ 2 2))
+'(+ 2 2)
+> (term ("hello" world (#false)))
+'("hello" world (#f))
+```
+
+Some terms may look strange.
+That's OK, because a term by itself has no meaning.
+
+
+### What is a _Redex model_?
+
+A Redex model is collection of tools for working with terms.
+The tools may include:
+
+- _languages_, to define a grammar for terms
+- _judgments_, to describe properties of terms or relations between terms
+- _metafunctions_, to transform terms
+- _reduction relations_, to defining a term rewriting system
+
+The goal of these tools is to encode a "real thing" (maybe, a programming language)
+ using Redex terms.
+
+
+### What is a language?
+
+A Redex _language_ is a named set of non-terminals, _patterns_, and _binding forms_.
+For example, a Redex model of the natural numbers might start with this language:
+
+```
+(define-language nat
+  [N ::= Zero
+         (Plus1 N)])
+```
+- the name of the language is `nat`
+- the non-terminal `N` is associated with two patterns: `Zero` and `(Plus1 N)`
+- there are no _binding forms_
+
+Each pattern describes a syntactic category of terms.
+Each non-terminal gives a name to the union of patterns that follow it.
+
+The non-terminal `N` describes all terms that are either:
+
+1. the symbol `Zero`
+2. lists of the form `(Plus1 N)`, where `N` is either `Zero` or another "Plus1"
+
+For example,
+
+```
+(term Zero)
+(term (Plus1 Zero))
+(term (Plus1 (Plus1 Zero)))
+(term (Plus1 (Plus1 (Plus1 Zero))))
+;; .... and so on
+```
+
+If a language has binding forms, then some terms can introduce names.
+See the FAQ entry on _binding forms_ for an example.
+
+
+### What is a pattern?
+
+A pattern is a sequence of characters and variables.
+If you have: (1) a language `L`, and (2) a pattern that contains _named non-terminals_ from `L`,
+ then you can ask whether a Redex term matches the pattern.
+
+If the term `t` matches the pattern, then Redex can show how the _named non-terminals_
+ in the pattern match sub-terms of `t`.
+
+For example, `(redex-match? L p t)` returns `#true` if the term `t` matches
+ the pattern `p` relative to the language `L`.
+
+```
+(define-language nat
+  [N ::= Zero (Plus1 N)])
+
+(redex-match? nat N_0 (term Zero))
```

Comment:
perhaps explain that the underscore suffix is special?

Reply:
I'll change the suffix, but I definitely don't want to drop it.
Because dropping suffixes has caused me many bugs and (I think) is confusing for beginners.